src/beliefs/discrete.jl

```diff
@@ -1,133 +1,107 @@
-# Maintained by Max Egorov and Zach Sunberg
+# Goals: minimize calls to ordered_states (allocates memory)
 
-"""
-A weight vector to be used as a belief or state distribution.
-
-sum(b)=1 is not enforced at all times, but during the update it is normalized to sum to one.
-
-pdf(b, i) calculates the sum every time it is called. To access the weight directly (for example if you are sure that the sum is 1), use weight(b, i). 
-"""
-mutable struct DiscreteBelief
+# needs pomdp for state_index in pdf(b, s)
+# needs list of ordered_states for rand(b)
+mutable struct DiscreteBelief{P<:POMDP, S}
+    pomdp::P
+    state_list::Vector{S}       # vector of ordered states
     b::Vector{Float64}
 end
+function DiscreteBelief(pomdp)
+    state_list = ordered_states(pomdp)
+    ns = n_states(pomdp)
 
-# Constructor with uniform belief
-DiscreteBelief(n::Int64) = DiscreteBelief(zeros(n) + 1.0/n)
+    b = ones(ns) / ns
 
-mutable struct DiscreteUpdater{P<:POMDP} <: Updater
-    pomdp::P
+    return DiscreteBelief(pomdp, state_list, b)
 end
 
-vec(b::DiscreteBelief) = b.b
 
-Base.length(b::DiscreteBelief) = length(b.b)
-index(b::DiscreteBelief, i::Int64) = i
-weight(b::DiscreteBelief, i::Int64) = b.b[i]
-iterator(b::DiscreteBelief) = filter(i->b[i]>0.0, 1:length(b))
-rand(rng::AbstractRNG, b::DiscreteBelief) = sample(rng, Weights(b.b)) # This will return an integer - seems like it should actually return an object of the state type of the problem
-pdf(b::DiscreteBelief, s::Int) = b.b[s]/sum(b.b) # only works when the state type is integer
+pdf(b::DiscreteBelief, s) = b.b[state_index(b.pomdp, s)]
+
+function rand(rng::AbstractRNG, b::DiscreteBelief)
+    i = sample(rng, Weights(b.b))
+    return b.state_list[i]
+end
 
 function Base.fill!(b::DiscreteBelief, x::Float64)
     fill!(b.b, x)
     return b
 end
 
-function Base.fill!(b::DiscreteBelief, idxs::Vector{Int64}, vals::Vector{Float64})
-    fill!(b.b, 0.0)
-    for i = 1:length(idxs)
-        index = idxs[i]
-        index > 0 ? (b[index] = vals[i]) : nothing 
+# should the belief just store the length so we don't have to count?
+Base.length(b::DiscreteBelief) = length(b.b)
+
+# I think this is a disaster and should be removed, but SARSOP.jl uses it
```

Comment:
I agree! If one of us doesn't fix it now, I think the best thing to do would be to file issues on SARSOP and POMDPToolbox that reference each other, then when someone decides to use SARSOP, they can fix it and have the satisfaction of contributing.

Reply:
I would even be ok with deprecating it now.

Reply:
It would be ok to deprecate now and file issues that reference each other. However, it hopefully won't last long. SARSOP uses the function with its alpha vector policy, and I think the next thing I want to work on is having an AlphaVectorSet policy so that QMDP, FIB, and SARSOP don't have to implement virtually the same types.